src/ocrd_network/runtime_data/hosts.py

```diff
@@ -0,0 +1,225 @@
+from logging import Logger
+from time import sleep
+from typing import Dict, List, Union
+
+from .connection_clients import create_docker_client, create_ssh_client
+from .network_agents import AgentType, DataNetworkAgent, DataProcessingWorker, DataProcessorServer, DeployType
+
+
+class DataHost:
+    def __init__(
+        self, host: str, username: str, password: str, keypath: str, workers: List[Dict], servers: List[Dict]
+    ) -> None:
+        self.host = host
+        self.username = username
+        self.password = password
+        self.keypath = keypath
+
+        # These flags are used to track whether a connection of the specified
+        # type should be created based on the received config file
+        self.needs_ssh_connector: bool = False
+        self.needs_docker_connector: bool = False
+
+        # Connection clients, ssh for native deployment, docker for docker deployment
+        self.ssh_client = None
+        self.docker_client = None
+
+        # Time to wait between deploying agents
+        self.wait_between_agent_deploys: float = 0.3
+
+        # Lists of network agents based on their agent and deployment type
+        self.network_agents_worker_native = []
+        self.network_agents_worker_docker = []
+        self.network_agents_server_native = []
+        self.network_agents_server_docker = []
+
+        if not workers:
+            workers = []
+        if not servers:
+            servers = []
+
+        self.__parse_network_agents_workers(processing_workers=workers)
+        self.__parse_network_agents_servers(processor_servers=servers)
+
+        # Used for caching deployed Processor Servers' ports on the current host
+        # Key: processor_name, Value: list of ports
+        self.processor_servers_ports: dict = {}
+
+    def __add_deployed_agent_server_port_to_cache(self, processor_name: str, port: int) -> None:
+        if processor_name not in self.processor_servers_ports:
+            self.processor_servers_ports[processor_name] = [port]
+            return
+        self.processor_servers_ports[processor_name] = self.processor_servers_ports[processor_name].append(port)
+
+    def __append_network_agent_to_lists(self, agent_data: DataNetworkAgent) -> None:
```

Comment:
Why the double underscores? If something starts with dunder `__`, I expect it to be some sort of official internal.

Reply:
To hint that it is an internal helper method and must not be called outside the class.

Reply:
I thought about the single and double leading underscores of function declarations too. Maybe I also was somebody who introduced them in some places. Previously I thought it was to indicate kind of private functions, but I googled a bit and came to the conclusion there is no one single way how this is understood. Tbh I think I would remove all these leading underscores (single and double) from "normal" functions and methods but I don't care too much.

Reply:
Not that important, as long it's not a "real" dunder method (ie. __foo__, with double underscores leading and trailing).